jmclient/jmclient/blockchaininterface.py

```diff
@@ -1,179 +1,308 @@
 
-    def fee_per_kb_has_been_manually_set(self, N):
-        '''if the 'block' target is higher than 1000, interpret it
+    @abstractmethod
+    def import_addresses(self, addr_list: List[str], wallet_name: str,
+                         restart_cb: Callable[[str], None] = None) -> None:
+        """Imports addresses in a batch during initial sync.
+        Refuses to proceed if keys are found to be under control
+        of another account/label (see console output), and quits.
+        """
+
+    @abstractmethod
+    def list_transactions(self, num: int, skip: int = 0) -> List[dict]:
+        """ Return a list of the last `num` transactions seen
+        in the wallet (under any label/account), optionally
+        skipping some.
+        """
+
+    @abstractmethod
+    def get_deser_from_gettransaction(self, rpcretval: dict) -> Optional[btc.CMutableTransaction]:
+        """Get full transaction deserialization from a call
+        to get_transaction().
+        """
+
+    @abstractmethod
+    def get_transaction(self, txid: bytes) -> Optional[dict]:
+        """ Argument txid is passed in binary.
+        Returns a serialized transaction for txid txid,
+        in hex as returned by Bitcoin Core rpc, or None
+        if no transaction can be retrieved. Works also for
+        watch-only wallets.
+        """
+
+    @abstractmethod
+    def get_block(self, blockheight: int) -> Optional[str]:
+        """Returns full hex serialized block at a given height.
+        """
+
+    @abstractmethod
+    def get_current_block_height(self) -> int:
+        """Returns the height of the most-work fully-validated chain.
+        """
+
+    @abstractmethod
+    def get_best_block_hash(self) -> str:
+        """Returns the hash of the best (tip) block in the most-work
+        fully-validated chain.
+        """
+
+    @abstractmethod
+    def get_best_block_median_time(self) -> int:
+        """Returns median time for the current best block.
+        """
+
+    @abstractmethod
+    def get_block_height(self, blockhash: str) -> int:
+        """Returns the block height for a specific block hash.
+        """
+
+    @abstractmethod
+    def get_block_time(self, blockhash: str) -> int:
+        """Returns the block time expressed in UNIX epoch time for a specific
+        block hash.
+        """
+
+    @abstractmethod
+    def get_block_hash(self, height: int) -> str:
+        """Returns hash of block in best-block-chain at height provided.
+        """
+
+    @abstractmethod
+    def get_tx_merkle_branch(self, txid: str,
+                             blockhash: Optional[str] = None) -> bytes:
+        """TODO: describe method.
+        """
+
+    @abstractmethod
+    def verify_tx_merkle_branch(self, txid: str, block_height: int,
+                                merkle_branch: bytes) -> bool:
+        """TODO: describe method.
+        """
+
+    @abstractmethod
+    def listaddressgroupings(self) -> list:
```

Comment:
In this case I feel like we're going too far in pulling methods out to the abstract class. The use of `listaddressgroupings` is very idiosyncratic and specific to the details of the bitcoind RPC interface.

Reply:
Yes, agree. But the same as above. We should look at changing calling code, then this can be removed from base class.

Reply:
Should I add FixMe or TODO comments for these maybe?

Reply:
I think this stuff is fine as per above comment that, if a method is not actually usable in a new implementation you can just define that method with `pass`. IIRC wallet service is explicitly using it, so if the clients are calling it, the interface currently does need it.

So, leave it for now is my guess.